### Here, three predictiing variables were considered to determine the varying prices of a house. At the end of the analysis, we would see which factor/factors (Area, distance, and crime rate) are insignificant  in the model.

In [131]:
# Import the libraries needed
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
import matplotlib.pyplot as plt

In [297]:
# import the dataset
dataset = pd.read_csv('/Users/lambertagunbiade/Desktop/SUPERVISED MACHINE LEARNING /landprice1.csv')
dataset

Area  Distance  Crime Rate  Price
0     89         4        3.84    7.0
1     66         1        3.19    5.4
2     78         3        3.78    6.6
3    111         6        3.89    7.4
4     44         1        3.57    4.8
5     77         3        3.57    6.4
6     80         3        3.03    7.0
7     66         2        3.51    5.6
8    109         5        3.54    7.3
9     75         3        3.25    6.4
10    85         2        3.31    6.9
11   110         6        3.20    7.3
12    45         1        3.70    4.8
13    79         1        3.21    6.8
14    90         4        3.87    7.0
15    76         2        3.56    6.5
16    81         3        3.21    7.0

* This data set shows two more variables unlike the previous data set but with same meaning.

#### Separate the target(price) variable from the predictors

In [133]:
X = dataset.iloc[0:,0:3].values # read all rows from index zero to the end but read only the columns between index 0 to 2
Y = dataset.iloc[0:, 3].values  # All rows and the last column(index number 3 only).

#### View the data sets again

In [134]:
print('X Predicting variables\n\n', X,'\n\nDependent variable Y \n\n',Y)

X Predicting variables

 [[ 89.     4.     3.84]
 [ 66.     1.     3.19]
 [ 78.     3.     3.78]
 [111.     6.     3.89]
 [ 44.     1.     3.57]
 [ 77.     3.     3.57]
 [ 80.     3.     3.03]
 [ 66.     2.     3.51]
 [109.     5.     3.54]
 [ 75.     3.     3.25]
 [ 85.     2.     3.31]
 [110.     6.     3.2 ]
 [ 45.     1.     3.7 ]
 [ 79.     1.     3.21]
 [ 90.     4.     3.87]
 [ 76.     2.     3.56]
 [ 81.     3.     3.21]] 

Dependent variable Y 

 [7.  5.4 6.6 7.4 4.8 6.4 7.  5.6 7.3 6.4 6.9 7.3 4.8 6.8 7.  6.5 7. ]


#### Split the data set into training and testing data

In [135]:
X_train,X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.4, random_state=0)

In [136]:
print(X_test)

[[ 66.     1.     3.19]
 [ 80.     3.     3.03]
 [109.     5.     3.54]
 [ 75.     3.     3.25]
 [ 79.     1.     3.21]
 [ 44.     1.     3.57]
 [ 78.     3.     3.78]]


#### Instantiate the linear regression class .

In [137]:
lin_regressor = LinearRegression()
lin_regressor.fit(X_train, Y_train)

LinearRegression()

### Model prediction


* Predict the price of each land with the area, crime rate and distance from the data set.

In [138]:
# Test data prediction
y_pred_test = lin_regressor.predict(X_test)

# Training data prediction
y_pred_train = lin_regressor.predict(X_train)

#### Testing data result

In [139]:
print('Original price of land\n\n',Y_test)

Original price of land

 [5.4 7.  7.3 6.4 6.8 4.8 6.6]


In [140]:
print('Price predicted by the model on test data\n', y_pred_test.round(2))

Price predicted by the model on test data
 [6.17 6.47 7.71 6.17 6.96 4.83 6.38]


* The result above shows that our model is predicting a little higher than the original prices of the land. Let us take a look at the trained data.


#### Training data result

In [141]:
print('Original prices on trained data',Y_train)

Original prices on trained data [7.  6.9 5.6 7.  7.3 7.4 7.  6.4 6.5 4.8]


In [142]:
print('predicted prices on trained data',y_pred_train.round(2))

predicted prices on trained data [6.84 7.06 5.9  6.53 7.48 7.57 6.78 6.31 6.52 4.9 ]


* The model performed better on the training data compare to the testing data in the sense that, the differences between the predicted prices are smaller compare to that of the test data. This would be easier to see from the value of R square of the training and test data respectively.

Let us see the value of R_Squared with all variables inclusive.

In [143]:
# Import r sqaured library and create an object of it thus;
from sklearn.metrics import r2_score

#### R_Sqaured value for the training datat

In [144]:
# R_Squared value for the trained and predicted trained values
r_square_train = r2_score(Y_train,y_pred_train)
r_square_train.round(2)

0.92

* The model is able to explain 92% variation in the training data set. However, we can not rely on this result as we must see how efficient it is on the test data.

#### R_Squared value for the test data.

In [145]:
# R_Squared value for the actual and predicted value
r_square_test = r2_score(Y_test,y_pred_test)
r_square_test.round(2)

0.76

• From the value of r squared, the model is able to explain only 76% variation in the test data set. However we can not rely on this for future predicitions because it has overfitted the training data.


#### Testing for multi collinearity

• Let us exclude "crime rate ", column to see the new values of r squared on train and test data respectively.•


In [146]:
X = dataset.iloc[0:,0:2].values # Read only the first two columns
Y = dataset.iloc[0:, 3].values # All rows and the last column(index number 3 only).

In [147]:
print('New values of X (Area and Distance)\n\n', X)

New values of X (Area and Distance)

 [[ 89   4]
 [ 66   1]
 [ 78   3]
 [111   6]
 [ 44   1]
 [ 77   3]
 [ 80   3]
 [ 66   2]
 [109   5]
 [ 75   3]
 [ 85   2]
 [110   6]
 [ 45   1]
 [ 79   1]
 [ 90   4]
 [ 76   2]
 [ 81   3]]


##### Split the data set into training and testing data

In [148]:
X_train,X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.4, random_state=0)

In [149]:
# Instantiate regression class
lin_regressor = LinearRegression()
lin_regressor.fit(X_train, Y_train)

LinearRegression()

### Now predict the model again with both training and testing data respectively


In [150]:
y_pred_test = lin_regressor.predict(X_test)
y_pred_train = lin_regressor.predict(X_train)

#### New R Squared values excluding "crime rate column".

In [151]:
# R_Squared value for the trained and predicted trained values
r_square_test = r2_score(Y_train,y_pred_train)
r_square_test.round(2)

0.92

In [152]:
# R_Squared value for the actual and predicted value
r_square_test = r2_score(Y_test,y_pred_test)
r_square_test.round(2)

0.76

 • It is obvious that the exlusion of crime rate from the dataset did not make any difference because the R squared value is same as when 'crime rate' column was used. This means that Distance and Crime rate columns, are highly correlated with each other.
 The two variables do not form a good relationship with price of the land ( The price of the land is independent of the crime rate in that neigbourhood ).
 
 From our simple linear regression on only the area of the land, we saw how the r squared values for both datasets had a close range (87% on the trained data and 81% on the test data).
 Let us demonstrate this by building a template on the three variables for easier understanding.

#### Now, let us build a template for any three variables.


In [153]:
# import the dataset

X = dataset.iloc[0:,0:3].values
Y = dataset.iloc[0:, 3].values

In [154]:
# Split the data into train and test data
X_train,X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.4, random_state=0)


# Fit the model
lin_regressor = LinearRegression()
lin_regressor.fit(X_train, Y_train)


# Model Prediction

# Test data prediction
y_pred_test = lin_regressor.predict(X_test)


#### R SQUARED VALUES

In [155]:
# R_SQUARED VALUE FOR TRAINING DATA
r_square_train = r2_score(Y_train,y_pred_train)
r_square_train.round(2)

0.92

In [156]:
# R_SQUARED VALUE FOR TEST DATA
r_square_test = r2_score(Y_test,y_pred_test)
r_square_test.round(2)

0.76

The model did performed excellently on the training data but not on the test data.

#### Now assume any value for Area, Distance and Crime rate to predict price.

In [157]:
transformed_values = np.array([[96,23,3.55]])
pred_price = lin_regressor.predict(transformed_values)
print(pred_price.round(2))

[1.94]


* Now let us consider same area of land at the city centre with more than twice the crime rate.

In [158]:
transformed_values = np.array([[96,0,9]])
pred_price = lin_regressor.predict(transformed_values)
print(pred_price.round(2))

[8.54]


* Considering the values inputed, the first has shown that the worth of a house on same size of land and 23 miles away from the city centre should be 1.94 million dollars. On the other hand, the second values inputed in the template says; same size of land at the city centre ( 0 distance value), with crime rate(9) more than twice that of an area of about 23 miles away from the city centre, is highly more valuable, attractive to buyers. Its demand far more than one which is far away. But in reality, is not absolutely true that an enviroment with high crime would not have negative effect on demand of houses. However, remember that crimes are of different phases. The exact kind of crime was not disclose in this case.

#### Now let us see the value of a house on same size of land with no crime or crime rate at its bearest minimum in same location of the city.

• Now, let us assume values of zero for other columns except the area of the land and see the price of the house.

In [159]:
transformed_values = np.array([[96,0,0]])
pred_price = lin_regressor.predict(transformed_values)
print(pred_price.round(2))

[8.14]


### Conclusion

* We have seen how much emphasis should be placed on the area or size of the land than other factors in considering the prices of a house in this region.

### LET US CONSIDER A DATASET WITH A CATEGORICAL VARIABLE

In [160]:
# Read the dataset
# import the dataset
dataset = pd.read_csv('/Users/lambertagunbiade/Desktop/SUPERVISED MACHINE LEARNING /landprice2.csv')
print('Data set preview\n\n', dataset)

Data set preview

     Area  Distance  Crime Rate  Price        City
0     89         4        3.84    7.0      Zurich
1     66         1        3.19    5.4      Jeneva
2     78         3        3.78    6.6  Interlaken
3    111         6        3.89    7.4      Jeneva
4     44         1        3.57    4.8      Zurich
5     77         3        3.57    6.4      Zurich
6     80         3        3.03    7.0  Interlaken
7     66         2        3.51    5.6      Jeneva
8    109         5        3.54    7.3      Zurich
9     75         3        3.25    6.4      Zurich
10    85         2        3.31    6.9  Interlaken
11   110         6        3.20    7.3  Interlaken
12    45         1        3.70    4.8      Jeneva
13    79         1        3.21    6.8  Interlaken
14    90         4        3.87    7.0      Jeneva
15    76         2        3.56    6.5      Jeneva
16    81         3        3.21    7.0      Zurich


#### Create dummy variable for the city column and preview

In [161]:
dummy = pd.get_dummies(dataset.City)
dummy

Interlaken  Jeneva  Zurich
0            0       0       1
1            0       1       0
2            1       0       0
3            0       1       0
4            0       0       1
5            0       0       1
6            1       0       0
7            0       1       0
8            0       0       1
9            0       0       1
10           1       0       0
11           1       0       0
12           0       1       0
13           1       0       0
14           0       1       0
15           0       1       0
16           0       0       1

 The dataframe shows three categories of data. However, we only need (n-1) number of dummy variables.
Let us select the first two columns with 'iloc' function.

In [162]:
dummy = pd.get_dummies(dataset.City).iloc[0: , 0:2]
dummy

Interlaken  Jeneva
0            0       0
1            0       1
2            1       0
3            0       1
4            0       0
5            0       0
6            1       0
7            0       1
8            0       0
9            0       0
10           1       0
11           1       0
12           0       1
13           1       0
14           0       1
15           0       1
16           0       0

• Explanation: We can clearly see that, if the city is neither of the two shown in the dataframe above, then it must be Zurich. Now we have only two dummy variables.
Next, we will concatenate the dummy variables to our dataset dataframe as below.

#### Concatenate both dataframes.

In [163]:
dataset= pd.concat([dataset,dummy], axis=1)   # axis = 1 because we are joining columns and not row.
dataset

Area  Distance  Crime Rate  Price        City  Interlaken  Jeneva
0     89         4        3.84    7.0      Zurich           0       0
1     66         1        3.19    5.4      Jeneva           0       1
2     78         3        3.78    6.6  Interlaken           1       0
3    111         6        3.89    7.4      Jeneva           0       1
4     44         1        3.57    4.8      Zurich           0       0
5     77         3        3.57    6.4      Zurich           0       0
6     80         3        3.03    7.0  Interlaken           1       0
7     66         2        3.51    5.6      Jeneva           0       1
8    109         5        3.54    7.3      Zurich           0       0
9     75         3        3.25    6.4      Zurich           0       0
10    85         2        3.31    6.9  Interlaken           1       0
11   110         6        3.20    7.3  Interlaken           1       0
12    45         1        3.70    4.8      Jeneva           0       1
13    79         1        3.21    6.8  Interlaken           1       0
14    90         4        3.87    7.0      Jeneva           0       1
15    76         2        3.56    6.5      Jeneva           0       1
16    81         3        3.21    7.0      Zurich           0       0

• Now we will have to drop the city column since its dummies have been created.

In [164]:
dataset = dataset.drop(['City'], axis=1)
dataset

Area  Distance  Crime Rate  Price  Interlaken  Jeneva
0     89         4        3.84    7.0           0       0
1     66         1        3.19    5.4           0       1
2     78         3        3.78    6.6           1       0
3    111         6        3.89    7.4           0       1
4     44         1        3.57    4.8           0       0
5     77         3        3.57    6.4           0       0
6     80         3        3.03    7.0           1       0
7     66         2        3.51    5.6           0       1
8    109         5        3.54    7.3           0       0
9     75         3        3.25    6.4           0       0
10    85         2        3.31    6.9           1       0
11   110         6        3.20    7.3           1       0
12    45         1        3.70    4.8           0       1
13    79         1        3.21    6.8           1       0
14    90         4        3.87    7.0           0       1
15    76         2        3.56    6.5           0       1
16    81         3        3.21    7.0           0       0

#### Rearrange the columns to make the target variable the last column

In [165]:
# Print the column variables
dataset.columns

Index(['Area', 'Distance', 'Crime Rate', 'Price', 'Interlaken', 'Jeneva'], dtype='object')

In [166]:
new_columns = ['Area', 'Distance', 'Crime Rate', 'Interlaken', 'Jeneva','Price']

In [167]:
dataset = dataset[new_columns]
dataset

Area  Distance  Crime Rate  Interlaken  Jeneva  Price
0     89         4        3.84           0       0    7.0
1     66         1        3.19           0       1    5.4
2     78         3        3.78           1       0    6.6
3    111         6        3.89           0       1    7.4
4     44         1        3.57           0       0    4.8
5     77         3        3.57           0       0    6.4
6     80         3        3.03           1       0    7.0
7     66         2        3.51           0       1    5.6
8    109         5        3.54           0       0    7.3
9     75         3        3.25           0       0    6.4
10    85         2        3.31           1       0    6.9
11   110         6        3.20           1       0    7.3
12    45         1        3.70           0       1    4.8
13    79         1        3.21           1       0    6.8
14    90         4        3.87           0       1    7.0
15    76         2        3.56           0       1    6.5
16    81         3        3.21           0       0    7.0

#### Separate the target(price) variable from the predictors

In [168]:
X = dataset.iloc[0:,0:5].values 
Y = dataset.iloc[0:, -1].values

In [169]:
print('X Predicting variables\n\n', X,'\n\nDependent variable Y \n\n',Y)

X Predicting variables

 [[ 89.     4.     3.84   0.     0.  ]
 [ 66.     1.     3.19   0.     1.  ]
 [ 78.     3.     3.78   1.     0.  ]
 [111.     6.     3.89   0.     1.  ]
 [ 44.     1.     3.57   0.     0.  ]
 [ 77.     3.     3.57   0.     0.  ]
 [ 80.     3.     3.03   1.     0.  ]
 [ 66.     2.     3.51   0.     1.  ]
 [109.     5.     3.54   0.     0.  ]
 [ 75.     3.     3.25   0.     0.  ]
 [ 85.     2.     3.31   1.     0.  ]
 [110.     6.     3.2    1.     0.  ]
 [ 45.     1.     3.7    0.     1.  ]
 [ 79.     1.     3.21   1.     0.  ]
 [ 90.     4.     3.87   0.     1.  ]
 [ 76.     2.     3.56   0.     1.  ]
 [ 81.     3.     3.21   0.     0.  ]] 

Dependent variable Y 

 [7.  5.4 6.6 7.4 4.8 6.4 7.  5.6 7.3 6.4 6.9 7.3 4.8 6.8 7.  6.5 7. ]


In [170]:
# Split the data set into training and testing data
X_train,X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.4, random_state=0)

#### Instantiate the linear regression class .

In [171]:
lin_regressor = LinearRegression()
lin_regressor.fit(X_train, Y_train)

LinearRegression()

#### MODEL PREDICTION

In [172]:
# Test data prediction
y_pred_test = lin_regressor.predict(X_test)

# Training data prediction
y_pred_train = lin_regressor.predict(X_train)

In [173]:
print('Original price of land\n\n',Y_test)
print('\n')
print('Price predicted by the model on test data\n', y_pred_test.round(2))

Original price of land

 [5.4 7.  7.3 6.4 6.8 4.8 6.6]


Price predicted by the model on test data
 [6.17 6.26 8.05 6.43 6.85 4.98 6.08]


We can see a big difference between the predicted prices and actual prices.

#### R SQUARED VALUES

In [174]:
# R_Squared value for the trained and predicted trained values
r_square_train = r2_score(Y_train,y_pred_train)
r_square_train.round(2)

0.97

In [175]:
# R_Squared value for the actual and predicted value
r_square_test = r2_score(Y_test,y_pred_test)
r_square_test.round(2)

0.59

From the R squared values, we can tell that the model has performed poorly on the test data but overfitted the training data.

#### Template to predict house prices in the three various cities of (Interlaken, Jeneva, Zurich)

### For Interlaken

In [176]:
transformed_values = np.array([[96,23,3.55,1,0]])
pred_price = lin_regressor.predict(transformed_values)
print(pred_price.round(2))

[0.69]


### For Jeneva

In [177]:
transformed_values = np.array([[96,23,3.55,0,1]])
pred_price = lin_regressor.predict(transformed_values)
print(pred_price.round(2))

[0.89]


### For Zurich

In [178]:
transformed_values = np.array([[96,23,3.55,0,0]])
pred_price = lin_regressor.predict(transformed_values)
print(pred_price.round(2))

[1.21]


If you look at the price prediction above, same values like we did previously were inputed considering same conditions in the Cities of Zurich, Interlaken and Jeneva.
Clearly, we can tell that despite the poor performance of our model, the cost of houses in Zurich are more expensive given same area, distance and crime rate. But we can not rely on this model because of its poor performance on the test data. Hence predicted prices are not reliable.
However, this massive gap between the training and testing data (model performance) can be checked with **regularization techniques(Ridge and Lasso regressions)**.
This will be seen from the **regularisation** section.

**This next section was better explained in the WEIGHT PREDICTION file.**

### REGULARIZATION

Regularization helps to solve the problem of overfitting and high variance. Also, to avoid model complexity by shrinking or reducing the values of the coefficients.
It simplifies a model by adding a penalty.

Regularization has two techniques: Ridge and Lasso regression.

##### Definition of some terms

Bias: 
    
    The difference between an actual values and average predictede values.
    High bias means high error in the model. This type of biasness does not work both on training and testing data
    
Variance:
    • Variance tells how the trained model will change on new testing data.
    High variance performs well on only training data and fail on test data. This is simply overfitting.

#### RIDGE REGRESSION

In [179]:
# import the ridge regression class
from sklearn.linear_model import Ridge

##### Instantiate the ridge regression class 

In [235]:
rid = Ridge(alpha = 30)
rid.fit(X_train, Y_train)

Ridge(alpha=30)

I observed that the value of R square of the training data reduced when the value of alpha increased but the decrease is not proportionate to the increase in the R Squared value of the testing data.

#### Model Prediction

In [236]:
# Training data prediction
y_pred_ridge_train = rid.predict(X_train)

# Testing data prediction
y_pred_ridge_test = rid.predict(X_test)

#### Compare the actual price vs the price predicted

In [237]:
print('Actual price\n\n', Y_test,'\n\n Predicted price\n\n', y_pred_ridge_test.round(2))

Actual price

 [5.4 7.  7.3 6.4 6.8 4.8 6.6] 

 Predicted price

 [5.95 6.48 7.61 6.28 6.5  5.07 6.4 ]


#### R Squared values

In [238]:
# Training data r square value
r_square_train = r2_score(Y_train, y_pred_ridge_train)

# Testing data r square value
r_square_test = r2_score(Y_test, y_pred_ridge_test)

print('R square values:- goodness of fit')
print('Training data: ', r_square_train.round(2),'Testing data: ',r_square_test.round(2) )

R square values:- goodness of fit
Training data:  0.88 Testing data:  0.82


### Using same values to predict houses prices in the three cities.

### For Interlaken

In [289]:
transformed_values = np.array([[96,23,3.55,1,0]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[6.58]


### For Jeneva

In [290]:
transformed_values = np.array([[96,23,3.55,0,1]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[6.57]


### For Zurich

In [291]:
transformed_values = np.array([[96,23,3.55,0,0]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[6.59]


In all of the cases in the choice of alpha, the difference in price between same size of land with other conditions, there is no much difference between one in Zurich and Interlaken.  

#### LASO REGRESSION

In [292]:
# import the ridge regression class
from sklearn.linear_model import Lasso

#### Instantiate the Lasso regression class and fit the model on the training data

In [334]:
lass = Lasso(alpha = 1) # Note: The choice of alpha is quite sensitive.
lass.fit(X_train, Y_train)

Lasso(alpha=1)

The choice of alpha = 1 gave the best perforce as the values were tested from 1 to 10. A higher value of alpha reduced the performance of the model in both training and testing data.

#### Model predicition

In [294]:
# Training data prediction
y_pred_lasso_train = lass.predict(X_train)

# Testing data prediction
y_pred_lasso_test = lass.predict(X_test)

#### R Squared values

In [295]:
# Training data r square value
lasso_r_square_train = r2_score(Y_train, y_pred_lasso_train)

# Testing data r square value
lasso_r_square_test = r2_score(Y_test, y_pred_lasso_test)
      
print('R square values:- goodness of fit')
print('Training data: ', lasso_r_square_train.round(2),'Testing data: ',lasso_r_square_test.round(2) )

R square values:- goodness of fit
Training data:  0.87 Testing data:  0.8


### Using same values to predict houses prices in the three cities.

### For Interlaken

In [296]:
transformed_values = np.array([[96,23,3.55,1,0]])
pred_price = lass.predict(transformed_values)
print(pred_price.round(2))

[7.06]


### For Jeneva

In [288]:
transformed_values = np.array([[96,23,3.55,0,1]])
pred_price = lass.predict(transformed_values)
print(pred_price.round(2))

[7.06]


### For Zurich

In [287]:
transformed_values = np.array([[96,23,3.55,0,0]])
pred_price = lass.predict(transformed_values)
print(pred_price.round(2))

[7.06]


This model says that, the prices in the prices of land in the three locations are the same.

### Let us see the first part of this analysis with three variables with any of the regularization models since the variables are few.

In [370]:
# import the dataset
dataset = pd.read_csv('/Users/lambertagunbiade/Desktop/SUPERVISED MACHINE LEARNING /landprice1.csv')
dataset

Area  Distance  Crime Rate  Price
0     89         4        3.84    7.0
1     66         1        3.19    5.4
2     78         3        3.78    6.6
3    111         6        3.89    7.4
4     44         1        3.57    4.8
5     77         3        3.57    6.4
6     80         3        3.03    7.0
7     66         2        3.51    5.6
8    109         5        3.54    7.3
9     75         3        3.25    6.4
10    85         2        3.31    6.9
11   110         6        3.20    7.3
12    45         1        3.70    4.8
13    79         1        3.21    6.8
14    90         4        3.87    7.0
15    76         2        3.56    6.5
16    81         3        3.21    7.0

#### Separate the predicting variables from the target variable.

In [374]:
X = dataset.iloc[0:,0:3].values 
Y = dataset.iloc[0:, 3].values 

#### Split the dataset into training and testing data.

In [375]:
X_train,X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.4, random_state=0)

#### Instatiate the Ridge regression class

In [376]:
rid = Ridge(alpha =15)
rid.fit(X_train, Y_train)

Ridge(alpha=15)

#### MODEL PREDICTION

In [377]:
# Training data prediction
y_pred_ridge_train = rid.predict(X_train)

# Testing data prediction
y_pred_ridge_test = rid.predict(X_test)

#### R SQUARED VALUES

In [378]:
# Training data r square value
r_square_train = r2_score(Y_train, y_pred_ridge_train)

# Testing data r square value
r_square_test = r2_score(Y_test, y_pred_ridge_test)

print('R square values:- goodness of fit')
print('Training data: ', r_square_train.round(2),'Testing data: ',r_square_test.round(2) )

R square values:- goodness of fit
Training data:  0.89 Testing data:  0.82


The model has reduced the overfitted training data performance by almost 10% and increased its performance on the testing data with about 4%.
This is reliable for future predictions. Let us see what would get with Lasso regression.

#### General template

**For a house at the city centre no crime rate recorded**

In [388]:
transformed_values = np.array([[96,0,0]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[7.35]


**For a house in crime free area and 50 miles away from the city centre**

In [389]:
transformed_values = np.array([[96,50,0]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[4.74]


**For a house at the city centre with high crime recorded.**

In [390]:
transformed_values = np.array([[96,0,9]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[7.29]


**This is more like a house in a country side with high crime rate.**

In [393]:
transformed_values = np.array([[96,100,24]])
pred_price = rid.predict(transformed_values)
print(pred_price.round(2))

[1.97]


The predictions of the models are looking great. It is true that a house in a crime free enviroment and closets to the city centre will be more expensive than otherwise.

### Laso Regression

#### fit the model on the training data

In [359]:
lass = Lasso(alpha = 1) # Note: The choice of alpha is quite sensitive.
lass.fit(X_train, Y_train)

Lasso(alpha=1)

An increase in the value of alpha has shown poor performance on the test data below 80% hence the choice of 1. Every other value greater than one reduced the performance of the model on the test data.

#### MODEL PREDICTION

In [360]:
# Training data prediction
y_pred_lasso_train = lass.predict(X_train)

# Testing data prediction
y_pred_lasso_test = lass.predict(X_test)

#### R SQUARED VALUES

In [361]:
# Training data r square value
lasso_r_square_train = r2_score(Y_train, y_pred_lasso_train)

# Testing data r square value
lasso_r_square_test = r2_score(Y_test, y_pred_lasso_test)
      
print('R square values:- goodness of fit')
print('Training data: ', lasso_r_square_train.round(2),'Testing data: ',lasso_r_square_test.round(2) )

R square values:- goodness of fit
Training data:  0.87 Testing data:  0.8


#### Conclusion

Although both models performed very well on both datasets with same percentage difference of 7. However, Ridge performed better on the test data while Lasso gave lower chances for overfitting. In conclusion, both models are reliable for future predictions.